### DISCUSSION TITLE
External data embeded in item api

### DISCUSSION THREAD

**Discussion by casualmatt** [directus/directus#17504] [Feature Requests]
### Summary

Fetching data from external services like Stripe, as it was a Json field in the db.

### Basic Example

In the data model page you could define a field that is actually a request to another services. 

```js
{
    "id": 1,
    "mail": "example@example.com",
    "name": "Example",
    "phone": "1234567890",
    "stripe_id": "cus_CE2Lygf3QSDzuY",
    "stripe": {
        //fetch
        // url: https://api.stripe.com/v1/customers/{{stripe_id}}
        // method: GET
        // headers: {...}
  }
  }


```

outcome: 
```js
{
  "id": 1,
  "mail": "example@example.com",
  "name": "Example",
  "phone": "1234567890",
  "stripe_id": "cus_CE2Lygf3QSDzuY",
  "stripe": {
    "id": "cus_CE2Lygf3QSDzuY",
    "object": "customer",
    "address": null,
    "balance": 0,
    "created": 1517223215
    // ...
  }
}


```

### Motivation

Now we need to relies on webhooks and directus flow with complicated sync procedure to get data from other sources, when we could just use use directus as a proxy to other backend. 

### Detailed Design

Adding a new field type with the ability to fetch data from external sources in real time. 
the interface for that field type could be something similar to this (just to give and idea).
![image](https://user-images.githubusercontent.com/85650530/218532816-8cf2edc1-905c-4576-944b-7a17d7750b99.png)

### Must Have:
- Simple interface, like a normal field
- Support for Rest API with basic or header auth
### Should Have:
- Possibility to use other field in the table as part of the URL
### Could Have:
- Possibility to use other field in the table as part of the Auth
- Possibility to execute a full directus flow with the output insert in the response object. 

### Won't Have:
- Complicated feature for that there is Flows and custom endpoints

### Drawbacks

Currently is already possible to do it, but we need to use custom endpoint to do that or directus flows, but they are both less intuitive to use when coding an app. 

### Alternatives

Using and external proxy to get the data to complete the information, still require a lot of time and privilege management can be tricky. 

### Adoption Strategy

Is not a breaking change, and it would allow developer to managed all there data from directus.

**Comment by rijkvanzanten** [comment]
Love this idea 🙂  I'd also like to think through if and how we could make this writable as well 🤔

**Reply by casualmatt** [reply]
in my mind i guest that after Knex return the data from the DB directus could check if they are 3° party data to fetch and include them in the body for the response. 

More in details, i'm not so familiar with directus jet to help, but i could take a look.

**Reply by rijkvanzanten** [reply]
Yeah, I think that's a bit too simplified unfortunately, as that'd mean that'd we'd have to ping Stripe `n` times, where `n` is the number of items returned from the API. If you were to fetch 1,000 records, it'd mean we'd have to fire a request to stripe once per parent, which is 1000 calls to Stripe's API. At their rate limit, that would already take 10 seconds, not even accounting for the time per request it takes to get the data. Do that with nested relational data, and that'll blow up 😬  To implement this at scale, it has to be way smarter at batching things by retrieving multiple items from a single call to the third party API to then stitch together with the database items. 🤔

**Reply by casualmatt** [reply]
I think it should be the developer to specify if he want that field, to be off by default and you need to strongly request it with `fields=stripe `

**Reply by casualmatt** [reply]
It could also be that it is not stripe but is just another directus or a custom server that you host and can manage 100 requests each time. 
In the end, it should be on the dev side to know a bit about the limitation of its own setup.

**Reply by casualmatt** [reply]
If you guys are interested in investigate how to do it, i will take some time to come out with a solution that use a mix of caching, tricks and rules that can make it feasible at least in for the API side.

**Reply by rijkvanzanten** [reply]
> In the end, it should be on the dev side to know a bit about the limitation of its own setup.

This really doesn't scale like you might expect when you have a wider community like Directus' 😬

> If you guys are interested in investigate how to do it, i will take some time to come out with a solution that use a mix of caching, tricks and rules that can make it feasible at least in for the API side.

It's an interesting use case for sure, but not something we're looking to support right away. Lets leave this feature request open for the time being to gauge overall community interest first! Implementing and maintaining such a feature long term is a non-trivial workload, so we'll have to make sure it's worth including in core at this point 🙂

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
